### PR TITLE
UID2-3401: Allow sending only v4 tokens to a specified list

### PR DESCRIPTION
**Pull request by caroline-ttd**

### CODE REVIEW

File: src/main/java/com/uid2/operator/service/UIDOperatorService.java [hunk 1]
```diff
@@ -311,11 +313,32 @@ private AdvertisingToken createAdvertisingToken(PublisherIdentity publisherIdent
             int hash = ((rawUid[0] & 0xFF) << 12) | ((rawUid[1] & 0xFF) << 4) | ((rawUid[2] & 0xFF) & 0xF); //using same logic as ModBasedSaltEntryIndexer.getIndex() in uid2-shared
             pseudoRandomNumber = (hash % 100) + 1; //1 to 100
         }
-
-        var tokenVersion = (pseudoRandomNumber <= this.advertisingTokenV4Percentage) ? TokenVersion.V4 : this.tokenVersionToUseIfNotV4;
+        TokenVersion tokenVersion;
+        if (siteIdsUsingV4Tokens.contains(publisherIdentity.siteId)) {
+            tokenVersion = TokenVersion.V4;
+        } else {
+            tokenVersion = (pseudoRandomNumber <= this.advertisingTokenV4Percentage) ? TokenVersion.V4 : this.tokenVersionToUseIfNotV4;
```
Comment (jon8787) [inline src/main/java/com/uid2/operator/service/UIDOperatorService.java:320]: `hash` and `pseudoRandomNumber` are not needed if this site is in `siteIdsUsingV4Tokens`. I'd suggest doing the site id check first with an early return.

File: src/test/java/com/uid2/operator/UIDOperatorServiceTest.java [hunk 2]
```diff
@@ -176,7 +177,64 @@ public void testGenerateAndRefresh() {
...
+        assertEquals(identityRequest.userIdentity.identityType, advertisingToken.userIdentity.identityType);
+        assertEquals(identityRequest.userIdentity.establishedAt, advertisingToken.userIdentity.establishedAt);
+
+        RefreshToken refreshToken = tokenEncoder.decodeRefreshToken(tokens.getRefreshToken());
+        assertEquals(this.now, refreshToken.createdAt);
+        assertEquals(this.now.plusSeconds(REFRESH_TOKEN_EXPIRES_AFTER_SECONDS), refreshToken.expiresAt);
+        assertEquals(identityRequest.publisherIdentity.siteId, refreshToken.publisherIdentity.siteId);
+        assertEquals(identityRequest.userIdentity.identityScope, refreshToken.userIdentity.identityScope);
+        assertEquals(identityRequest.userIdentity.identityType, refreshToken.userIdentity.identityType);
+        assertEquals(identityRequest.userIdentity.establishedAt, refreshToken.userIdentity.establishedAt);
```
Comment (jon8787) [inline src/test/java/com/uid2/operator/UIDOperatorServiceTest.java:224]: there are a few nearly duplicate blocks of assertion code in this MR - can we refactor into a method?

File: src/test/java/com/uid2/operator/UIDOperatorServiceTest.java [hunk 3]
```diff
@@ -176,7 +177,64 @@ public void testGenerateAndRefresh() {
...
+
+        setNow(Instant.now().plusSeconds(200));
+
+        reset(shutdownHandler);
```
Comment (jon8787) [inline src/test/java/com/uid2/operator/UIDOperatorServiceTest.java:228]: likewise - is this code duplicated, is it possible to refactor into a method?

File: src/test/java/com/uid2/operator/UIDOperatorServiceTest.java [hunk 4]
```diff
@@ -176,7 +177,64 @@ public void testGenerateAndRefresh() {
...
+        assertEquals(refreshToken.userIdentity.identityScope, refreshToken2.userIdentity.identityScope);
+        assertEquals(refreshToken.userIdentity.identityType, refreshToken2.userIdentity.identityType);
+        assertEquals(refreshToken.userIdentity.establishedAt, refreshToken2.userIdentity.establishedAt);
+        assertArrayEquals(refreshToken.userIdentity.id, refreshToken2.userIdentity.id);
+    }
+
+    @Test
+    public void testGenerateAndRefreshForSiteIdUsingV4Token() {
+        final IdentityRequest identityRequest = new IdentityRequest(
+                new PublisherIdentity(127, 124, 125),
```
Comment (sunnywu) [inline src/test/java/com/uid2/operator/UIDOperatorServiceTest.java:201]: could you parameterize this test so same test is run for both site id 127 & 128 as you set them both for site_ids_using_v4_tokens?

File: src/main/java/com/uid2/operator/service/UIDOperatorService.java [hunk 5]
```diff
@@ -311,11 +313,32 @@ private AdvertisingToken createAdvertisingToken(PublisherIdentity publisherIdent
...
+        }
         return new AdvertisingToken(tokenVersion, now, now.plusMillis(identityExpiresAfter.toMillis()), this.operatorIdentity, publisherIdentity, userIdentity);
     }
 
+    private Set<Integer> getSiteIdsUsingV4Tokens(String siteIdsUsingV4TokensInString) {
```
Comment (sunnywu) [inline src/main/java/com/uid2/operator/service/UIDOperatorService.java:325]: can this be a static method in some helper class/file outside UIDOperatorService? 
can you add a few simple unit tests for it? (inputs like empty string, invalid string , 1 site id, multiple site ids without spaces everywhere and in between that should be trimmed)

**Review by sunnywu — APPROVED** [review]

File: src/main/java/com/uid2/operator/service/UIDOperatorService.java (hunk 1 above)
Comment (caroline-ttd) [reply src/main/java/com/uid2/operator/service/UIDOperatorService.java:321]: Yeah, that makes sense. I moved them to the else block.

File: src/test/java/com/uid2/operator/UIDOperatorServiceTest.java (hunk 3 above)
Comment (caroline-ttd) [reply src/test/java/com/uid2/operator/UIDOperatorServiceTest.java]: I refactored my test cases using `testGenerateAndRefresh`

File: src/test/java/com/uid2/operator/UIDOperatorServiceTest.java [hunk 6]
```diff
@@ -176,7 +177,64 @@ public void testGenerateAndRefresh() {
...
+        assertEquals(this.now.plusSeconds(REFRESH_TOKEN_EXPIRES_AFTER_SECONDS), refreshToken2.expiresAt);
+        assertEquals(refreshToken.publisherIdentity.siteId, refreshToken2.publisherIdentity.siteId);
+        assertEquals(refreshToken.userIdentity.identityScope, refreshToken2.userIdentity.identityScope);
+        assertEquals(refreshToken.userIdentity.identityType, refreshToken2.userIdentity.identityType);
+        assertEquals(refreshToken.userIdentity.establishedAt, refreshToken2.userIdentity.establishedAt);
+        assertArrayEquals(refreshToken.userIdentity.id, refreshToken2.userIdentity.id);
+    }
+
+    @Test
+    public void testGenerateAndRefreshForSiteIdUsingV4Token() {
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/test/java/com/uid2/operator/UIDOperatorServiceTest.java]: I added a comments "site id 127 and 128 is for testing "site_ids_using_v4_tokens"

File: src/test/java/com/uid2/operator/UIDOperatorServiceTest.java (hunk 4 above)
Comment (caroline-ttd) [reply src/test/java/com/uid2/operator/UIDOperatorServiceTest.java]: Sure, I added the 128 site id test case too.

File: src/main/java/com/uid2/operator/service/UIDOperatorService.java (hunk 5 above)
Comment (caroline-ttd) [reply src/main/java/com/uid2/operator/service/UIDOperatorService.java]: I moved the `getSiteIdsUsingV4Tokens` class to TokenUtils and created correspond tests in `TokenUtilsTest.class`

File: src/test/java/com/uid2/operator/UIDOperatorServiceTest.java (hunk 2 above)
Comment (caroline-ttd) [reply src/test/java/com/uid2/operator/UIDOperatorServiceTest.java]:
```
assertEquals(identityRequest.userIdentity.identityScope, refreshToken.userIdentity.identityScope);
assertEquals(identityRequest.userIdentity.identityType, refreshToken.userIdentity.identityType);
assertEquals(identityRequest.userIdentity.establishedAt, refreshToken.userIdentity.establishedAt); 
```
As the code block is using the same input, I created a private method `assertIdentityScopeIdentityTypeAndEstablishedAt(UserIdentity expctedUserIdentity, UserIdentity actualUserIdentity)`

**Review by jon8787 — APPROVED** [review]

File: src/test/java/com/uid2/operator/UIDOperatorServiceTest.java [hunk 7]
```diff
@@ -176,21 +180,18 @@ public void testGenerateAndRefresh() {
...
         assertArrayEquals(advertisingToken.userIdentity.id, advertisingToken2.userIdentity.id);
 
         RefreshToken refreshToken2 = tokenEncoder.decodeRefreshToken(refreshResponse.getTokens().getRefreshToken());
         assertEquals(this.now, refreshToken2.createdAt);
         assertEquals(this.now.plusSeconds(REFRESH_TOKEN_EXPIRES_AFTER_SECONDS), refreshToken2.expiresAt);
         assertEquals(refreshToken.publisherIdentity.siteId, refreshToken2.publisherIdentity.siteId);
-        assertEquals(refreshToken.userIdentity.identityScope, refreshToken2.userIdentity.identityScope);
-        assertEquals(refreshToken.userIdentity.identityType, refreshToken2.userIdentity.identityType);
-        assertEquals(refreshToken.userIdentity.establishedAt, refreshToken2.userIdentity.establishedAt);
+        assertIdentityScopeIdentityTypeAndEstablishedAt(refreshToken.userIdentity, refreshToken2.userIdentity);
```
[the remark this answers is not in the document]
Comment (caroline-ttd) [reply src/test/java/com/uid2/operator/UIDOperatorServiceTest.java:194]: If we include `publisherIdentity` as a parameter, there will be four parameters in the method `assertIdentityScopeIdentityTypeAndEstablishedAt`, and only one line of code will be needed to assert the site ID for publisherIdentity in each block. I feel it isn't worthwhile to include it unless there are additional assertions for publisherIdentity. 

I am open to other suggestions and happy to make the change if needed.

[the remark this answers is not in the document]
Comment (jon8787) [reply src/test/java/com/uid2/operator/UIDOperatorServiceTest.java:194]: ok
